src/PowerShellEditorServices/Analysis/AnalysisService.cs

```diff
@@ -115,8 +116,8 @@ public AnalysisService(string settingsPath, ILogger logger)
                 this.SettingsPath = settingsPath;
 
                 scriptAnalyzerModuleInfo = FindPSScriptAnalyzerModule(logger);
-                var sessionState = InitialSessionState.CreateDefault2();
-                sessionState.ImportPSModulesFromPath(scriptAnalyzerModuleInfo.ModuleBase);
+                InitialSessionState sessionState = InitialSessionState.CreateDefault2();
+                sessionState.ImportPSModulesFromPath(GetModulePathFromModuleInfo(scriptAnalyzerModuleInfo));
```

Comment:
Since we're just grabbing PSSA from the module path, I wonder if there's a helper variable somewhere that already has the Module path in it. Then we don't need  to do this fancy path logic with PSSA just to get the module path.

Reply:
Yeah that's a good idea. It may even be a matter of querying the variable.

Reply:
I've been thinking about this. It looks like the way to access this directly is [kept internal](https://github.com/PowerShell/PowerShell/blob/2a45b48ae2f05c007636d716e41a9c60df6488ec/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs#L1009). It's present as a property of `ExecutionContext`, but getting that out of the current runspace would be tricky and possibly error prone.

So another option is to get our powershell session to evaluate `$env:PSModulePath`, but I don't think that's very performant. My bigger concern is that, even if we don't consider the possibility that the module path could be changed in some way we don't expect, we still would be trying to import every module in the PSES module directory. Which we don't need in the analysis part and could potentially be quite slow.

But it's also true that I've overengineered the module path resolution bit already. Since we sort of know specifically whether PSSA will be using a versioned path or not. Except that this broke before. And ideally we could eventually move this method out and make it a static utility function.

Reply:
Keep in mind PSSA can come from anywhere, we definitely don't want to lock it to the bundled version.

Reply:
Ah, that's a very good point (which admittedly I had forgotten)!

But I'm pretty sure the new code still does that. It still retrieves the module given by `Get-Module PSScriptAnalyzer` and gets the path. The problem is that `PSModuleInfo` doesn't have a good way of representing its base path; `PSModuleInfo.ModuleBase` gives you what could be a versioned path that doesn't work with `Import-Module` (or the C# API equivalent used here).

So the only bit I added was to remove the versioned directory if it's the leaf directory.

Reply:
Oh for sure, I should have clarified that I was commenting on the thread and not the actual changes.  Side note, I usually just target the `psd1` in these situations.  Something like this

```csharp
Path.ChangeExtension(
    Path.Combine(
        moduleInfo.ModuleBase,
        moduleInfo.Name),
    ".psd1"))
```

But your approach accounts for modules without a manifest so it's more flexible in general.

Reply:
Well actually I wanted to do that (or use whatever `PSModuleInfo.Path` gave me. Or there's also an overload of `InitialSessionState.ImportPSModulesFromPath` that takes an `IEnumerable<ModuleSpecification>`, and there's a constructor for `ModuleSpecification` that takes a `PSModuleInfo`.

**But**, it turns out that `ImportPSModulesFromPath` calls `ModuleUtils.GetDefaultAvailableModuleFiles` internally, [which only accepts paths to directories](https://github.com/PowerShell/PowerShell/blob/2a45b48ae2f05c007636d716e41a9c60df6488ec/src/System.Management.Automation/engine/Modules/ModuleUtils.cs#L200) as arguments.

(And the second method doesn't seem to populate the `ModuleSpecification` fields other than `Name`.)

I just wish PowerShell exposed a more convenient way to manipulate its modules and module loading.

Reply:
Ah right, I always forget the differences between those methods. I typically use `InitialSessionState.ImportPSModule`, which does work with `psd1` files.

> I just wish PowerShell exposed a more convenient way to manipulate its modules and module loading.

👍 👍 👍

Reply:
In a quest to work out what might be best, I noticed that `InitialSessionState.ImportPSModulesFromPath()` in turn just converts the modules it finds into `ModuleSpecification`s and feeds them to `InitialSessionState.ImportPSModule`.

Anyway, I went to look at why this particular command can take a `ModuleSpecification` with everything but the `Name` field set to `null`, and have the `Name` be just a name or a path to a specification, or a whole heap of other things, and just work.

Well, turns out no matter what you do, PowerShell itself just calls `Import-Module` (see [here](https://github.com/PowerShell/PowerShell/blob/38c53c50a623f7c54707d348dd42d68a583b8932/src/System.Management.Automation/engine/InitialSessionState.cs#L3015)).

So I might actually just change our code to do that directly (even though I don't really like the idea of that)...